### PR TITLE
[BC BREAK] `MessageBag` does not depend on `\ArrayObject` anymore

### PR DESCRIPTION
**Pull request by OskarStark**
Follows
* https://github.com/php-llm/llm-chain/pull/166/files#r1898145794

### CODE REVIEW

File: src/Bridge/Meta/LlamaPromptConverter.php [hunk 1]
```diff
@@ -19,7 +19,7 @@ public function convertToPrompt(MessageBag $messageBag): string
         $messages = [];
 
         /** @var UserMessage|SystemMessage|AssistantMessage $message */
-        foreach ($messageBag->getIterator() as $message) {
+        foreach ($messageBag->messages as $message) {
```
Comment (OskarStark) [inline src/Bridge/Meta/LlamaPromptConverter.php:22]: I like the `getIterator()` more

File: tests/Model/Message/MessageBagTest.php [hunk 2]
```diff
@@ -70,7 +70,7 @@ public function with(): void
         self::assertCount(3, $messageBag);
         self::assertCount(4, $newMessageBag);
 
-        $newMessageFromBag = $newMessageBag[3];
+        $newMessageFromBag = $newMessageBag->messages[3];
```
Comment (OskarStark) [inline tests/Model/Message/MessageBagTest.php:73]: I like the previous version more TBH

File: src/Model/Message/MessageBag.php [hunk 3]
```diff
@@ -4,19 +4,26 @@
...
-/**
- * @template-extends \ArrayObject<int, MessageInterface>
- */
-final class MessageBag extends \ArrayObject implements \JsonSerializable
+final class MessageBag implements \Countable, \JsonSerializable
 {
+    /**
+     * @var MessageInterface[]
+     */
+    public array $messages = [];
```
Comment (chr-hertel) [inline src/Model/Message/MessageBag.php:12]: if we directly expose the array we cannot ensure any type safety here. everyone could (def including me) skip the `add(...)` method and just write on the array. no matter if `MessageInterface` or not.

too scared, would prefer to be more defensive here and have a `getMessages()` instead

File: src/Model/Message/MessageBag.php [hunk 4]
```diff
@@ -4,19 +4,26 @@
...
+
     public function __construct(MessageInterface ...$messages)
     {
-        parent::__construct(array_values($messages));
+        $this->messages = array_values($messages);
+    }
+
+    public function add(MessageInterface $message): void
```
Comment (chr-hertel) [inline src/Model/Message/MessageBag.php:19]: what about
```suggestion
    public function add(MessageInterface ...$messages): void
```
to support adding multiple messages?

File: src/Model/Message/MessageBag.php [hunk 5]
```diff
@@ -73,11 +78,16 @@ public function containsImage(): bool
         return false;
     }
 
+    public function count(): int
+    {
+        return count($this->messages);
+    }
```
Comment (chr-hertel) [inline src/Model/Message/MessageBag.php:84]: i like :+1:

**Comment by chr-hertel** [comment]
@OskarStark looks like you have mixed feelings due to missing iterator and array access?

i'm quite positive about this tbh

File: src/Model/Message/MessageBag.php (hunk 3 above)
Comment (OskarStark) [reply src/Model/Message/MessageBag.php]: done

**Comment by OskarStark** [comment]
Now with the `getMessages()` its more clear as an API, so I am fine 👍

File: src/Model/Message/MessageBag.php (hunk 4 above)
Comment (OskarStark) [reply src/Model/Message/MessageBag.php:19]: Hmm the use case is quite unusual I would say.

**Review by chr-hertel — APPROVED** [review]
LGTM

**Pull request merged**